x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/SearchableSnapshotAction.java

```diff
@@ -112,57 +151,124 @@ boolean isForceMergeIndex() {
+                    return false;
                 }
+                // Perform the searchable snapshot, as the index is not currently a searchable snapshot
                 return false;
             });
-        CheckNotDataStreamWriteIndexStep checkNoWriteIndexStep = new CheckNotDataStreamWriteIndexStep(checkNoWriteIndex,
-            waitForNoFollowerStepKey);
-        final WaitForNoFollowersStep waitForNoFollowersStep;
-        if (forceMergeIndex) {
-            waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, forceMergeStepKey, client);
-        } else {
-            waitForNoFollowersStep = new WaitForNoFollowersStep(waitForNoFollowerStepKey, generateSnapshotNameKey, client);
-        }
+        CheckNotDataStreamWriteIndexStep checkNoWriteIndexStep =
+            new CheckNotDataStreamWriteIndexStep(checkNoWriteIndex, waitForNoFollowerStepKey);
+        WaitForNoFollowersStep waitForNoFollowersStep =
+            new WaitForNoFollowersStep(waitForNoFollowerStepKey, skipGeneratingSnapshotKey, client);
+
+        // When generating a snapshot, we either jump to the force merge step, or we skip the
+        // forcemerge and go straight to steps for creating the snapshot
+        StepKey keyForSnapshotGeneration = forceMergeIndex ? forceMergeStepKey : generateSnapshotNameKey;
+        // Branch, deciding whether there is an existing searchable snapshot snapshot that can be used for mounting the index
+        // (in which case, skip generating a new name and the snapshot cleanup), or if we need to generate a new snapshot
+        BranchingStep skipGeneratingSnapshotStep =
+            new BranchingStep(skipGeneratingSnapshotKey, keyForSnapshotGeneration, mountSnapshotKey, (index, clusterState) -> {
+                IndexMetadata indexMetadata = clusterState.getMetadata().index(index);
+                if (indexMetadata == null) {
+                    return false;
+                }
+                String policyName = LifecycleSettings.LIFECYCLE_NAME_SETTING.get(indexMetadata.getSettings());
+                LifecycleExecutionState lifecycleExecutionState = LifecycleExecutionState.fromIndexMetadata(indexMetadata);
+                if (lifecycleExecutionState.getSnapshotName() == null) {
+                    // No name exists, so it must be generated
+                    logger.trace("no snapshot name for index [{}] in policy [{}] exists, so one will be generated",
+                        index.getName(), policyName);
+                    return false;
+                }
+
+                if (this.snapshotRepository.equals(lifecycleExecutionState.getSnapshotRepository()) == false) {
+                    // A different repository is being used
+                    // TODO: allow this behavior instead of throwing an exception
+                    throw new IllegalArgumentException("searchable snapshot indices may be converted only within the same repository");
```

Comment:
shall we prevent this at policy validation time for now?

Reply:
Not yet I think, I'm going to try to get it done this week, and if I don't have time, I'll add the policy validation at the end of the week (trying to avoid adding the policy validation and then immediately removing it)